sycl/include/sycl/reduction.hpp

```diff
@@ -755,17 +897,31 @@ class reduction_impl_algo : public reduction_impl_common<T, BinaryOperation> {
       });
     };
     if constexpr (is_usm) {
-      // Don't dispatch based on base::initializeToIdentity() as that would lead
+      // Don't dispatch based on initializeToIdentity() as that would lead
       // to two different instantiations of Func.
       DoIt(MRedOut);
     } else {
-      if (base::initializeToIdentity())
+      if (initializeToIdentity())
         DoIt(MRedOut);
       else
         Func(MRedOut);
     }
   }
 
+  // Overload of withInitializedMem for reducer without identity. Initializing
+  // to identity is not allowed in this case.
+  template <typename KernelName, typename FuncTy,
+            bool HasIdentity = has_identity>
+  std::enable_if_t<!HasIdentity> withInitializedMem(handler &CGH, FuncTy Func) {
+    assert(!initializeToIdentity() &&
+           "Initialize to identity not allowed for identity-less reductions.");
+    Func(MRedOut);
+  }
```

Comment:
I'd very much prefer `if constexpr` for that purpose.

Reply:
Since we would need both branches of the conditional, we would be indenting a lot of code from the above definition of the function, so personally I prefer this solution. If you feel very strongly about it though, I am open to changing it.

Reply:
Yeah, I agree. Didn't notice the size of the "other" branch.